### PR TITLE
Yolo multilabel

### PR DESCRIPTION
**Pull request by Danil328**
### YOLO fair multilabel.
**NOT FOR MERGE**

The idea was as follows:
1. The dataset format unchanged (label, cx, cy, w, h):
```
0 0.43335 0.154441 0.114746 0.308883
4 0.43335 0.154441 0.114746 0.308883
10 0.43335 0.154441 0.114746 0.308883
17 0.43335 0.154441 0.114746 0.308883
18 0.43335 0.154441 0.114746 0.308883
0 0.393555 0.580754 0.183594 0.26245
4 0.393555 0.580754 0.183594 0.26245
15 0.393555 0.580754 0.183594 0.26245
17 0.393555 0.580754 0.183594 0.26245
18 0.393555 0.580754 0.183594 0.26245
```
2. When loading the dataset, we find unique boxes and aggregate the labels into OHE format.
3. Change the loss, tal, metrics, plotting for multilabel support.

Training starts, metrics converge, the only thing is that the model prediction after NSM from the multi-label format is turned back into the multi-class format.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR adds native support for multi-label object detection in Ultralytics YOLO, allowing each object to have multiple class labels instead of just one. 🏷️🚀

### 📊 Key Changes
- **Multi-label Option Added:** New `multilabel` setting in the config file to enable multi-label detection.
- **New Dataset Class:** Introduces `YOLOMultilabelDataset` to handle multi-label data during training and validation.
- **Data Pipeline Updates:** Adjustments throughout the data loading, augmentation, and batching processes to support multi-label annotations.
- **Model & Loss Adjustments:** Modifies model training, loss calculation, and post-processing to work with multi-label targets.
- **Metrics & Visualization:** Updates metrics calculations and plotting functions to correctly handle and display multi-label results.
- **Backward Compatibility:** Maintains standard single-label detection as default; multi-label is opt-in.

### 🎯 Purpose & Impact
- **Broader Use Cases:** Enables training and evaluating models where objects can belong to multiple categories at once (e.g., "person" and "athlete").
- **Improved Flexibility:** Users can now tackle more complex datasets and real-world scenarios with overlapping class definitions.
- **Seamless Integration:** Multi-label support is integrated into the existing workflow, making it easy to switch on when needed.
- **Enhanced Analytics:** Metrics and visualizations now accurately reflect multi-label performance, giving users deeper insights.

This update empowers users to build more versatile and powerful detection models with YOLO! 💡🖼️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Danil328, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! Here's a quick checklist to help ensure the seamless integration of your contribution:

- ✅ **Clearly Define the PR Purpose**: In your PR description, please ensure the purpose is clearly outlined and, if possible, link to any [relevant issues or feature requests](https://github.com/ultralytics/ultralytics/issues). It’s helpful to summarize how this enhancement benefits the project. 
- ✅ **Synchronize with Source**: Ensure your branch is up-to-date with the latest `main` branch of the [Ultralytics repository](https://github.com/ultralytics/ultralytics). If needed, update your branch by clicking 'Update branch' or running `git pull` and `git merge main` locally.
- ✅ **Pass CI/CD Tests**: Confirm that all [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks pass successfully. If any fail, please address the accompanying errors.
- ✅ **Update Documentation**: Update any relevant [documentation](https://docs.ultralytics.com) associated with the changes, such as YAML configuration descriptions, new arguments, or method usage.
- ✅ **Add Tests & Validation**: If applicable, add tests to validate your multilabel functionality and include outputs to demonstrate its effectiveness.
- ✅ **Sign the CLA**: Ensure you have signed our [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/). If you haven’t already, please comment "I have read the CLA Document and I sign the CLA" in this PR thread.
- ✅ **Minimize Changes**: While adding significant new functionality like multilabel handling is impressive 🎉, strive to keep changes concise and focused on your feature to reduce the overhead during review.

For further guidance, please check out the [Ultralytics Contributing Guide](https://docs.ultralytics.com/help/contributing).

Please note, this is an automated response 🛠️. An Ultralytics engineer will carefully review your PR soon. If you have additional questions or require clarification, feel free to comment below. Thank you for contributing to Ultralytics! 🚀

**Comment by Danil328** [comment]
I have read the CLA Document and I sign the CLA

**Comment by hu874** [comment]
> ### YOLO fair multilabel.
> **NOT FOR MERGE**
> 
> The idea was as follows:
> 
> 1. The dataset format unchanged (label, cx, cy, w, h):
> 
> ```
> 0 0.43335 0.154441 0.114746 0.308883
> 4 0.43335 0.154441 0.114746 0.308883
> 10 0.43335 0.154441 0.114746 0.308883
> 17 0.43335 0.154441 0.114746 0.308883
> 18 0.43335 0.154441 0.114746 0.308883
> 0 0.393555 0.580754 0.183594 0.26245
> 4 0.393555 0.580754 0.183594 0.26245
> 15 0.393555 0.580754 0.183594 0.26245
> 17 0.393555 0.580754 0.183594 0.26245
> 18 0.393555 0.580754 0.183594 0.26245
> ```
> 
> 2. When loading the dataset, we find unique boxes and aggregate the labels into OHE format.
> 3. Change the loss, tal, metrics, plotting for multilabel support.
> 
> Training starts, metrics converge, the only thing is that the model prediction after NSM from the multi-label format is turned back into the multi-class format.
> 
> ## 🛠️ PR Summary
> Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)
> 
> ### 🌟 Summary
> This PR introduces support for multi-label object detection in YOLO, enhancing the flexibility and accuracy of the model in handling datasets where objects may belong to multiple classes simultaneously.
> 
> ### 📊 Key Changes
> * **New Multi-Label Setting**:
>   
>   * Added a `multilabel` option in the configuration file (`default.yaml`) to enable multi-label classification.
> * **New Dataset Class**:
>   
>   * Introduced `YOLOMultilabelDataset` to handle multi-label datasets effectively.
> * **Improvements to Data Processing**:
>   
>   * Updated data loading and augmentation to allow multi-label handling.
> * **Metrics and Loss Adjustments**:
>   
>   * Revised metrics and loss functions to appropriately handle multi-label targets, including precision-recall calculations and heatmap generation.
> * **Visualization Updates**:
>   
>   * Adjusted plots and visualizations to reflect multi-label annotations correctly.
> 
> ### 🎯 Purpose & Impact
> * **Purpose**:
>   
>   * Enable the YOLO model to support multi-label datasets, addressing a significant limitation in scenarios where objects can belong to multiple categories.
> * **Impact**:
>   
>   * Greater versatility for real-world applications like medical imaging or retail, where multi-label classifications are common.
>   * Improved model flexibility without disrupting the functionalities of single-label or traditional YOLO tasks.
>   * Users can now seamlessly train on multi-label datasets, unlocking new use cases and expanding YOLO's adaptability. 🛠️💡

Is there any progress on this work? Can it be used? grateful

**Comment by glenn-jocher** [comment]
Thank you for your interest in this work! As the PR is marked **NOT FOR MERGE**, it is currently not part of the official Ultralytics repository and cannot be directly used. However, you can monitor any updates or discussions on the PR in the repository. If multi-label support is officially integrated in the future, it will be announced in the [release notes](https://github.com/ultralytics/ultralytics/releases).

**Comment by KungFuPandaPro** [comment]
could you provide your code and implement details?

**Comment by glenn-jocher** [comment]
Thank you for your interest! Please refer to the PR description and associated files for implementation details. If you'd like to explore or test the approach, you can clone the branch and review the code directly. Let us know if you have specific questions!
